### PR TITLE
[EXP][CMD-BUFFER] Add read and write function definitions

### PR DESCRIPTION
**Pull request by mfrancepillois**
Add new command-buffer functions for appending read and write commands for one-dimensional and rectangular buffers

### CODE REVIEW

File: scripts/core/exp-command-buffer.yml [hunk 1]
```diff
@@ -259,6 +259,84 @@ returns:
...
+    - type: $x_mem_handle_t
+      name: hBuffer
+      desc: "[in] handle of the buffer object."
+    - type: "size_t"
+      name: offset
+      desc: "[in] offset in bytes in the buffer object."
+    - type: "size_t"
+      name: size
+      desc: "[in] size in bytes of data being written."
+    - type: "const void *"
```
Comment (kbenzie) [inline scripts/core/exp-command-buffer.yml:278]: the generator scripts don't handle a space before `*` well, I believe that's why the github actions failed.

```suggestion
    - type: "const void*"
```

File: scripts/core/exp-command-buffer.yml [hunk 2]
```diff
@@ -259,6 +259,84 @@ returns:
...
+    - type: $x_mem_handle_t
+      name: hBuffer
+      desc: "[in] handle of the buffer object."
+    - type: "size_t"
+      name: offset
+      desc: "[in] offset in bytes in the buffer object."
+    - type: "size_t"
+      name: size
+      desc: "[in] size in bytes of data being written."
+    - type: "void *"
```
Comment (kbenzie) [inline scripts/core/exp-command-buffer.yml:317]:
```suggestion
    - type: "void*"
```

File: scripts/core/exp-command-buffer.yml [hunk 3]
```diff
@@ -313,6 +391,114 @@ returns:
...
+    - type: "size_t"
+      name: bufferSlicePitch
+      desc: "[in] [in] length of each 2D slice in bytes in the buffer object being written."
+    - type: "size_t"
+      name: hostRowPitch
+      desc: "[in] length of each row in bytes in the host memory region pointed by dst."
+    - type: "size_t"
+      name: hostSlicePitch
+      desc: "[in] length of each 2D slice in bytes in the host memory region pointed by dst."
+    - type: "void *"
```
Comment (kbenzie) [inline scripts/core/exp-command-buffer.yml:425]:
```suggestion
    - type: "void*"
```

File: scripts/core/exp-command-buffer.yml [hunk 4]
```diff
@@ -313,6 +391,114 @@ returns:
...
+    - type: "size_t"
+      name: bufferSlicePitch
+      desc: "[in] [in] length of each 2D slice in bytes in the buffer object being read."
+    - type: "size_t"
+      name: hostRowPitch
+      desc: "[in] length of each row in bytes in the host memory region pointed by dst."
+    - type: "size_t"
+      name: hostSlicePitch
+      desc: "[in] length of each 2D slice in bytes in the host memory region pointed by dst."
+    - type: "void *"
```
Comment (kbenzie) [inline scripts/core/exp-command-buffer.yml:479]:
```suggestion
    - type: "void*"
```

**Review by kbenzie — CHANGES_REQUESTED** [review]

File: scripts/core/exp-command-buffer.yml [hunk 5]
```diff
@@ -313,6 +391,114 @@ returns:
...
+      desc: "[in] 3D rectangular region descriptor: width, height, depth."
+    - type: "size_t"
+      name: bufferRowPitch
+      desc: "[in] length of each row in bytes in the buffer object."
+    - type: "size_t"
+      name: bufferSlicePitch
+      desc: "[in] [in] length of each 2D slice in bytes in the buffer object being written."
+    - type: "size_t"
+      name: hostRowPitch
+      desc: "[in] length of each row in bytes in the host memory region pointed by dst."
```
Comment (Bensuo) [inline scripts/core/exp-command-buffer.yml:421]:
```suggestion
      desc: "[in] length of each row in bytes in the host memory region pointed to by pSrc."
```

File: scripts/core/exp-command-buffer.yml [hunk 6]
```diff
@@ -313,6 +391,114 @@ returns:
...
+    - type: "size_t"
+      name: hostSlicePitch
+      desc: "[in] length of each 2D slice in bytes in the host memory region pointed by dst."
```
Comment (Bensuo) [inline scripts/core/exp-command-buffer.yml:424]:
```suggestion
      desc: "[in] length of each 2D slice in bytes in the host memory region pointed to by pSrc."
```

File: scripts/core/exp-command-buffer.yml [hunk 7]
```diff
@@ -313,6 +391,114 @@ returns:
...
+      desc: "[in] 3D offset in the host region."
+    - type: $x_rect_region_t
+      name: region
+      desc: "[in] 3D rectangular region descriptor: width, height, depth."
+    - type: "size_t"
+      name: bufferRowPitch
+      desc: "[in] length of each row in bytes in the buffer object."
+    - type: "size_t"
+      name: bufferSlicePitch
+      desc: "[in] [in] length of each 2D slice in bytes in the buffer object being read."
```
Comment (Bensuo) [inline scripts/core/exp-command-buffer.yml:472]:
```suggestion
      desc: "[in] length of each 2D slice in bytes in the buffer object being read."
```

File: scripts/core/exp-command-buffer.yml [hunk 8]
```diff
@@ -313,6 +391,114 @@ returns:
...
+    - type: "size_t"
+      name: hostRowPitch
+      desc: "[in] length of each row in bytes in the host memory region pointed by dst."
```
Comment (Bensuo) [inline scripts/core/exp-command-buffer.yml:475]:
```suggestion
      desc: "[in] length of each row in bytes in the host memory region pointed to by pDst."
```

File: scripts/core/exp-command-buffer.yml [hunk 9]
```diff
@@ -313,6 +391,114 @@ returns:
...
+    - type: "size_t"
+      name: hostSlicePitch
+      desc: "[in] length of each 2D slice in bytes in the host memory region pointed by dst."
```
Comment (Bensuo) [inline scripts/core/exp-command-buffer.yml:478]:
```suggestion
      desc: "[in] length of each 2D slice in bytes in the host memory region pointed to by pDst."
```

**Review by Bensuo — COMMENTED** [review]
LGTM, just some minor comments

**Review by kbenzie — COMMENTED** [review]
Does `scripts/core/EXP-COMMAND-BUFFER.rst` need updated?

File: scripts/core/EXP-COMMAND-BUFFER.rst [hunk 10]
```diff
@@ -188,11 +201,12 @@ Changelog
 +-----------+------------------------+
 | Revision  | Changes                |
 +===========+========================+
-| 1.0       | Initial Draft           |
+| 1.1       | Initial Draft          |
```
Comment (Bensuo) [inline scripts/core/EXP-COMMAND-BUFFER.rst:204]: We should have this as a new line in the table, keeping the `1.0` row, and list what was changed in this version.

**Review by Bensuo — APPROVED** [review]
LGTM, just one minor comment!

**Review by Bensuo — APPROVED** [review]

File: scripts/core/registry.yml [hunk 11]
```diff
@@ -367,9 +367,18 @@ etors:
 - name: COMMAND_BUFFER_FINALIZE_EXP
   desc: Enumerator for $xCommandBufferFinalizeExp
   value: '123'
+- name: COMMAND_BUFFER_APPEND_MEMBUFFER_WRITE_EXP
```
Comment (veselypeta) [inline scripts/core/registry.yml:370]: Annoyingly this shouldn't happen - we should always allocate new registry entries from the largest value. Let me investigate what's going on here. :+1:

**Review by veselypeta — CHANGES_REQUESTED** [review]
LGMT - just blocking this to fix the registry.

File: scripts/core/registry.yml [hunk 12]
```diff
@@ -367,9 +367,18 @@ etors:
...
+  desc: Enumerator for $xCommandBufferAppendMembufferWriteExp
+  value: '124'
```
Comment (kbenzie) [inline scripts/core/registry.yml:372]: Looks like these ID's were generated before we updated the script to only append after the highest ID value.

If you could undo all the changes in this file, then rerun the `generate` target and push the new version I'd appreciate it. All of these entry-points *should* be grouped together at the end of the file.

**Review by kbenzie — CHANGES_REQUESTED** [review]
One last change, otherwise looking good!

**Review by veselypeta — APPROVED** [review]

File: scripts/core/registry.yml (hunk 11 above)
Comment (mfrancepillois) [reply scripts/core/registry.yml]: I did that and it solves the issue. Thanks

File: scripts/core/registry.yml (hunk 12 above)
Comment (mfrancepillois) [reply scripts/core/registry.yml]: Thanks. The re-generation solved the issue. :+1:

**Review by kbenzie — APPROVED** [review]

**Pull request merged**